# NumPy Mastery: 300 Comprehensive Questions
**Retail Dataset Analysis**

**Name:** ___________________

**Date:** ___________________

---

## Overview
This notebook contains **300 NumPy questions** divided into three difficulty levels:
- **Beginner (100 questions):** Fundamentals and basic operations
- **Intermediate (100 questions):** Advanced operations and techniques
- **Advanced (100 questions):** Complex problem-solving and optimization

## Dataset
We'll work with a **Retail Transaction Dataset** containing:
- 10,000 transactions
- Customer information
- Product details
- Sales data
- Timestamps

---

## Setup: Create Retail Dataset

Run this cell first to create our retail dataset that we'll use throughout the notebook.

In [1]:
import numpy as np
import pandas as pd
from datetime import datetime, timedelta

# Set random seed for reproducibility
np.random.seed(42)

# Generate retail dataset
n_transactions = 10000
n_products = 100
n_customers = 500
n_stores = 10

# Create the dataset
retail_data = {
    'transaction_id': np.arange(1, n_transactions + 1),
    'date': pd.date_range(start='2023-01-01', periods=n_transactions, freq='H'),
    'customer_id': np.random.randint(1, n_customers + 1, n_transactions),
    'product_id': np.random.randint(1, n_products + 1, n_transactions),
    'store_id': np.random.randint(1, n_stores + 1, n_transactions),
    'quantity': np.random.randint(1, 10, n_transactions),
    'unit_price': np.round(np.random.uniform(5, 500, n_transactions), 2),
    'discount': np.random.choice([0, 0.05, 0.10, 0.15, 0.20], n_transactions),
    'payment_method': np.random.choice(['Credit Card', 'Cash', 'Debit Card', 'Mobile Pay'], n_transactions),
    'customer_age': np.random.randint(18, 75, n_transactions),
    'customer_gender': np.random.choice(['M', 'F', 'Other'], n_transactions)
}

df = pd.DataFrame(retail_data)
df['total_amount'] = df['quantity'] * df['unit_price'] * (1 - df['discount'])
df['month'] = df['date'].dt.month
df['day_of_week'] = df['date'].dt.dayofweek
df['hour'] = df['date'].dt.hour

print("Retail Dataset Shape:", df.shape)
print("\nFirst few rows:")
print(df.head())
print("\nDataset Info:")
print(df.info())
print("\nBasic Statistics:")
print(df.describe())

# Save dataset for later use
df.to_csv('retail_data.csv', index=False)
print("\nDataset saved as 'retail_data.csv'")


C:\Users\user\AppData\Local\Temp\ipykernel_11704\2148428357.py:17: FutureWarning: 'H' is deprecated and will be removed in a future version, please use 'h' instead.
  'date': pd.date_range(start='2023-01-01', periods=n_transactions, freq='H'),


Retail Dataset Shape: (10000, 15)

First few rows:
   transaction_id                date  customer_id  product_id  store_id  \
0               1 2023-01-01 00:00:00          103          54        10   
1               2 2023-01-01 01:00:00          436          13         6   
2               3 2023-01-01 02:00:00          349          83         1   
3               4 2023-01-01 03:00:00          271          29        10   
4               5 2023-01-01 04:00:00          107           3         9   

   quantity  unit_price  discount payment_method  customer_age  \
0         9      305.21      0.05    Credit Card            29   
1         3      168.88      0.15           Cash            58   
2         2      211.88      0.15     Debit Card            44   
3         4      205.83      0.20     Debit Card            36   
4         4       47.88      0.15    Credit Card            74   

  customer_gender  total_amount  month  day_of_week  hour  
0               F     2609.5455    

---

# PART 1: BEGINNER LEVEL (100 Questions)

## Section 1.1: Array Basics (Questions 1-20)

### Question 1: Load Dataset as NumPy Array (Easy)

Load the retail dataset CSV file and convert it to a NumPy array. Extract only the numeric columns.

**Hint:** Use pd.read_csv() then convert to numpy array with .values or .to_numpy()

In [4]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df.to_numpy()

array([[1, '2023-01-01 00:00:00', 103, ..., 1, 6, 0],
       [2, '2023-01-01 01:00:00', 436, ..., 1, 6, 1],
       [3, '2023-01-01 02:00:00', 349, ..., 1, 6, 2],
       ...,
       [9998, '2024-02-21 13:00:00', 5, ..., 2, 2, 13],
       [9999, '2024-02-21 14:00:00', 410, ..., 2, 2, 14],
       [10000, '2024-02-21 15:00:00', 414, ..., 2, 2, 15]],
      shape=(10000, 15), dtype=object)

### Question 2: Extract Transaction IDs (Easy)

Extract all transaction_id values as a 1D NumPy array.

**Hint:** Use array slicing or column selection

In [ ]:
# Your code here
import pandas as pd
df = pd.read_csv('retail_data.csv')
transaction_ids = df['transaction_id'].to_numpy()
print(transaction_ids.shape)   # should be (n_transactions,)
print(transaction_ids[:10])    # preview first 10



(10000,)
[ 1  2  3  4  5  6  7  8  9 10]


### Question 3: Get Array Shape (Easy)

Find the shape, size, and dimensions of the retail data array.

**Hint:** Use .shape, .size, .ndim

In [6]:
# Your code here
print(df.shape)
print(df.size)
print(df.ndim)


(10000, 15)
150000
2


### Question 4: Data Types (Easy)

Check the data type of each column in the array.

**Hint:** Use .dtype attribute

In [7]:
# Your code here
df.dtypes


transaction_id       int64
date                object
customer_id          int64
product_id           int64
store_id             int64
quantity             int64
unit_price         float64
discount           float64
payment_method      object
customer_age         int64
customer_gender     object
total_amount       float64
month                int64
day_of_week          int64
hour                 int64
dtype: object

### Question 5: Create Quantity Array (Easy)

Create a NumPy array containing only the quantity column.

**Hint:** Index or slice the specific column

In [13]:
# Your code here
quantity_array_2d = df[['quantity']].to_numpy()
quantity_array_2d


array([[9],
       [3],
       [2],
       ...,
       [9],
       [3],
       [3]], shape=(10000, 1))

### Question 6: Create Price Array (Easy)

Extract unit_price values as a separate array.

**Hint:** Similar to previous

In [15]:
# Your code here
unit_price=df['unit_price'].to_numpy()
unit_price


array([305.21, 168.88, 211.88, ..., 218.62, 288.31, 163.93],
      shape=(10000,))

### Question 7: First 10 Transactions (Easy)

Get the first 10 rows of data.

**Hint:** Use array slicing [:10]

In [23]:
# Your code here
df.iloc[:10,:]


transaction_id                 date  customer_id  product_id  store_id  \
0               1  2023-01-01 00:00:00          103          54        10   
1               2  2023-01-01 01:00:00          436          13         6   
2               3  2023-01-01 02:00:00          349          83         1   
3               4  2023-01-01 03:00:00          271          29        10   
4               5  2023-01-01 04:00:00          107           3         9   
5               6  2023-01-01 05:00:00           72          90         7   
6               7  2023-01-01 06:00:00          189          58         7   
7               8  2023-01-01 07:00:00           21          76         7   
8               9  2023-01-01 08:00:00          103           2         2   
9              10  2023-01-01 09:00:00          122          75         2   

   quantity  unit_price  discount payment_method  customer_age  \
0         9      305.21      0.05    Credit Card            29   
1         3      168.88      0.15           Cash            58   
2         2      211.88      0.15     Debit Card            44   
3         4      205.83      0.20     Debit Card            36   
4         4       47.88      0.15    Credit Card            74   
5         3        6.77      0.10     Mobile Pay            43   
6         2      449.72      0.05    Credit Card            71   
7         5      112.81      0.00     Debit Card            32   
8         4      215.91      0.15     Mobile Pay            48   
9         5      486.10      0.00    Credit Card            63   

  customer_gender  total_amount  month  day_of_week  hour  
0               F     2609.5455      1            6     0  
1               F      430.6440      1            6     1  
2               M      360.1960      1            6     2  
3               M      658.6560      1            6     3  
4               M      162.7920      1            6     4  
5               M       18.2790      1            6     5  
6               F      854.4680      1            6     6  
7               F      564.0500      1            6     7  
8               M      734.0940      1            6     8  
9               M     2430.5000      1            6     9

### Question 8: Last 5 Transactions (Easy)

Get the last 5 transactions.

**Hint:** Use negative indexing [-5:]

In [33]:
# Your code here
df.iloc[-5:,:]

transaction_id                 date  customer_id  product_id  store_id  \
9995            9996  2024-02-21 11:00:00           57          15         9   
9996            9997  2024-02-21 12:00:00          112          36         8   
9997            9998  2024-02-21 13:00:00            5          47         2   
9998            9999  2024-02-21 14:00:00          410          54         7   
9999           10000  2024-02-21 15:00:00          414          86         1   

      quantity  unit_price  discount payment_method  customer_age  \
9995         4      213.91      0.05           Cash            54   
9996         7      406.74      0.05           Cash            53   
9997         9      218.62      0.00     Mobile Pay            66   
9998         3      288.31      0.10     Mobile Pay            62   
9999         3      163.93      0.15     Mobile Pay            72   

     customer_gender  total_amount  month  day_of_week  hour  
9995               M      812.8580      2            2    11  
9996               F     2704.8210      2            2    12  
9997               M     1967.5800      2            2    13  
9998               F      778.4370      2            2    14  
9999               F      418.0215      2            2    15

### Question 9: Every 100th Transaction (Easy)

Extract every 100th transaction from the dataset.

**Hint:** Use step in slicing [::100]

In [35]:
# Your code here
df.iloc[100:101,:]

transaction_id                 date  customer_id  product_id  store_id  \
100             101  2023-01-05 04:00:00          446          38         9   

     quantity  unit_price  discount payment_method  customer_age  \
100         6      303.09      0.15     Debit Card            30   

    customer_gender  total_amount  month  day_of_week  hour  
100               M      1545.759      1            3     4

### Question 10: Specific Columns (Easy)

Extract transaction_id, quantity, and unit_price columns.

**Hint:** Use column indexing

In [37]:
# Your code here
df[['transaction_id','quantity','unit_price']].to_numpy

<bound method DataFrame.to_numpy of       transaction_id  quantity  unit_price
0                  1         9      305.21
1                  2         3      168.88
2                  3         2      211.88
3                  4         4      205.83
4                  5         4       47.88
...              ...       ...         ...
9995            9996         4      213.91
9996            9997         7      406.74
9997            9998         9      218.62
9998            9999         3      288.31
9999           10000         3      163.93

[10000 rows x 3 columns]>

### Question 11: Array Reshaping (Easy)

Reshape the quantity array into a 100x100 matrix.

**Hint:** Use .reshape(100, 100)

In [39]:
# Your code here
quantity_array = df['quantity'].to_numpy()
quantity_matrix = quantity_array.reshape(100, 100)


### Question 12: Flatten Array (Easy)

Take a 2D subset and flatten it to 1D.

**Hint:** Use .flatten() or .ravel()

In [41]:
# Your code here
quantity_array = df['quantity'].to_numpy()
quantity_array.ravel()

array([9, 3, 2, ..., 9, 3, 3], shape=(10000,))

### Question 13: Create Zero Array (Easy)

Create an array of zeros with the same shape as quantities.

**Hint:** Use np.zeros_like()

In [ ]:
# Your code here
quantity_array = df['quantity'].to_numpy()
np.zeros_like('quantity_array')


array('', dtype='<U14')

### Question 14: Create Ones Array (Easy)

Create an array of ones matching the shape of prices.

**Hint:** Use np.ones_like()

In [45]:
# Your code here
quantity_array = df['quantity'].to_numpy()
np.ones_like('quantity_array')

array('1', dtype='<U14')

### Question 15: Create Range Array (Easy)

Create an array from 0 to 999 representing transaction indices.

**Hint:** Use np.arange()

In [ ]:
# Your code here
import numpy as np
transaction_indices = np.arange(0, 1000)
print(transaction_indices.shape)  # (1000,)
print(transaction_indices[:10])   # first 10 values


(1000,)
[0 1 2 3 4 5 6 7 8 9]


### Question 16: Create Linspace (Easy)

Create 50 evenly spaced values between min and max price.

**Hint:** Use np.linspace()

In [3]:
# Your code here
import numpy as np
print(np.linspace(1,30,5))


[ 1.    8.25 15.5  22.75 30.  ]


### Question 17: Copy Array (Easy)

Create a deep copy of the quantity array.

**Hint:** Use .copy()

In [ ]:
# Your code here


### Question 18: Array Memory (Easy)

Check the memory usage of the retail data array.

**Hint:** Use .nbytes

In [55]:
# Your code here
quantity_array = df['quantity'].to_numpy()
print(quantity_array.nbytes)

80000


### Question 19: Transpose Array (Easy)

Transpose a subset of the retail data (first 100 rows, 5 columns).

**Hint:** Use .T or .transpose()

In [51]:
# Your code here
df.T


0                    1     \
transaction_id                     1                    2   
date             2023-01-01 00:00:00  2023-01-01 01:00:00   
customer_id                      103                  436   
product_id                        54                   13   
store_id                          10                    6   
quantity                           9                    3   
unit_price                    305.21               168.88   
discount                        0.05                 0.15   
payment_method           Credit Card                 Cash   
customer_age                      29                   58   
customer_gender                    F                    F   
total_amount               2609.5455              430.644   
month                              1                    1   
day_of_week                        6                    6   
hour                               0                    1   

                                2                    3     \
transaction_id                     3                    4   
date             2023-01-01 02:00:00  2023-01-01 03:00:00   
customer_id                      349                  271   
product_id                        83                   29   
store_id                           1                   10   
quantity                           2                    4   
unit_price                    211.88               205.83   
discount                        0.15                  0.2   
payment_method            Debit Card           Debit Card   
customer_age                      44                   36   
customer_gender                    M                    M   
total_amount                 360.196              658.656   
month                              1                    1   
day_of_week                        6                    6   
hour                               2                    3   

                                4                    5     \
transaction_id                     5                    6   
date             2023-01-01 04:00:00  2023-01-01 05:00:00   
customer_id                      107                   72   
product_id                         3                   90   
store_id                           9                    7   
quantity                           4                    3   
unit_price                     47.88                 6.77   
discount                        0.15                  0.1   
payment_method           Credit Card           Mobile Pay   
customer_age                      74                   43   
customer_gender                    M                    M   
total_amount                 162.792               18.279   
month                              1                    1   
day_of_week                        6                    6   
hour                               4                    5   

                                6                    7     \
transaction_id                     7                    8   
date             2023-01-01 06:00:00  2023-01-01 07:00:00   
customer_id                      189                   21   
product_id                        58                   76   
store_id                           7                    7   
quantity                           2                    5   
unit_price                    449.72               112.81   
discount                        0.05                  0.0   
payment_method           Credit Card           Debit Card   
customer_age                      71                   32   
customer_gender                    F                    F   
total_amount                 854.468               564.05   
month                              1                    1   
day_of_week                        6                    6   
hour                               6                    7   

                                8                    9     ...  \
transaction_id                     9                   10 

### Question 20: Stacking Arrays (Easy)

Stack quantity and price arrays horizontally.

**Hint:** Use np.hstack() or np.column_stack()

In [ ]:


quantity_array = df['quantity'].to_numpy()
price_array = df['unit_price'].to_numpy()

# Stack horizontally (2D array: columns = quantity, unit_price)
stacked_array = np.hstack((quantity_array, price_array))
stacked_array


array([  9.  ,   3.  ,   2.  , ..., 218.62, 288.31, 163.93],
      shape=(20000,))

## Section 1.2: Basic Mathematical Operations (Questions 21-40)

### Question 21: Total Revenue (Easy)

Calculate total revenue by multiplying quantity by unit_price for all transactions.

**Hint:** Use element-wise multiplication

In [4]:
# Your code here
unit_price=4
transactions=5
total_revenue=unit_price*transactions
print(total_revenue)


20


### Question 22: Average Quantity (Easy)

Calculate the mean quantity sold.

**Hint:** Use np.mean()

In [6]:
# Your code here
import numpy as np
a=[1,2,3,4]
print(np.mean(a))


2.5


### Question 23: Total Quantity Sold (Easy)

Sum all quantities in the dataset.

**Hint:** Use np.sum()

In [13]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df

print(np.sum(df["quantity"]))


50539


### Question 24: Maximum Price (Easy)

Find the highest unit price.

**Hint:** Use np.max()

In [15]:
# Your code here
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
print(np.max(df["unit_price"]))

499.9


### Question 25: Minimum Quantity (Easy)

Find the smallest quantity sold.

**Hint:** Use np.min()

In [18]:
# Your code here
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
print(np.min(df["quantity"]))


1


### Question 26: Price Range (Easy)

Calculate the difference between max and min prices.

**Hint:** max - min

In [22]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
import numpy as np
print(np.max(df["total_amount"])-np.min(df["total_amount"]))


4484.5055


### Question 27: Square of Quantities (Easy)

Calculate the square of each quantity value.

**Hint:** Use ** 2 or np.square()

In [19]:
# Your code here
import numpy as np
print(np.square(df["quantity"]))

0       81
1        9
2        4
3       16
4       16
        ..
9995    16
9996    49
9997    81
9998     9
9999     9
Name: quantity, Length: 10000, dtype: int64


### Question 28: Square Root of Prices (Easy)

Calculate square root of all unit prices.

**Hint:** Use np.sqrt()

In [25]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
print(np.sqrt(df["unit_price"]))


0       17.470260
1       12.995384
2       14.556098
3       14.346777
4        6.919538
          ...    
9995    14.625662
9996    20.167796
9997    14.785804
9998    16.979694
9999    12.803515
Name: unit_price, Length: 10000, dtype: float64


### Question 29: Round Prices (Easy)

Round all prices to nearest integer.

**Hint:** Use np.round()

In [26]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
print(np.round(df["unit_price"]))



0       305.0
1       169.0
2       212.0
3       206.0
4        48.0
        ...  
9995    214.0
9996    407.0
9997    219.0
9998    288.0
9999    164.0
Name: unit_price, Length: 10000, dtype: float64


### Question 30: Absolute Values (Easy)

Calculate absolute values (relevant if there were negative values).

**Hint:** Use np.abs()

In [28]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
print(np.abs(df["unit_price"]))



0       305.21
1       168.88
2       211.88
3       205.83
4        47.88
         ...  
9995    213.91
9996    406.74
9997    218.62
9998    288.31
9999    163.93
Name: unit_price, Length: 10000, dtype: float64


### Question 31: Cumulative Sum (Easy)

Calculate cumulative sum of quantities.

**Hint:** Use np.cumsum()

In [29]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
print(np.cumsum(df["quantity"]))


0           9
1          12
2          14
3          18
4          22
        ...  
9995    50517
9996    50524
9997    50533
9998    50536
9999    50539
Name: quantity, Length: 10000, dtype: int64


### Question 32: Cumulative Product (Easy)

Calculate cumulative product of first 10 quantities.

**Hint:** Use np.cumprod()

In [30]:
# Your code here
import numpy as np
cum_prod = np.cumprod(df["quantity"].head(10))
print(cum_prod)



0         9
1        27
2        54
3       216
4       864
5      2592
6      5184
7     25920
8    103680
9    518400
Name: quantity, dtype: int64


### Question 33: Percentage Calculation (Easy)

Calculate what percentage each transaction is of total revenue.

**Hint:** Divide by sum and multiply by 100

In [ ]:
# Your code here


### Question 34: Price After Discount (Easy)

Calculate final price after applying discount.

**Hint:** price * (1 - discount)

In [ ]:
# Your code here


### Question 35: Revenue per Transaction (Easy)

Calculate revenue for each transaction (quantity * unit_price).

**Hint:** Element-wise multiplication

In [ ]:
# Your code here


### Question 36: Add Constant (Easy)

Add a $5 handling fee to all unit prices.

**Hint:** Use + operator

In [ ]:
# Your code here


### Question 37: Multiply by Constant (Easy)

Apply a 10% price increase to all unit prices.

**Hint:** Multiply by 1.10

In [ ]:
# Your code here


### Question 38: Exponential (Easy)

Calculate e raised to the power of (quantity/10).

**Hint:** Use np.exp()

In [ ]:
# Your code here


### Question 39: Logarithm (Easy)

Calculate natural log of all unit prices.

**Hint:** Use np.log()

In [ ]:
# Your code here


### Question 40: Power Operation (Easy)

Raise prices to the power of 1.5.

**Hint:** Use np.power()

In [ ]:
# Your code here


## Section 1.3: Basic Statistical Operations (Questions 41-60)

### Question 41: Mean Price (Easy)

Calculate average unit price.

**Hint:** np.mean()

In [ ]:
# Your code here


### Question 42: Median Quantity (Easy)

Find median quantity sold.

**Hint:** np.median()

In [ ]:
# Your code here


### Question 43: Standard Deviation (Easy)

Calculate standard deviation of prices.

**Hint:** np.std()

In [ ]:
# Your code here


### Question 44: Variance (Easy)

Calculate variance of quantities.

**Hint:** np.var()

In [ ]:
# Your code here


### Question 45: Min and Max (Easy)

Find both min and max prices in one operation.

**Hint:** np.min(), np.max()

In [ ]:
# Your code here


### Question 46: Percentiles (Easy)

Calculate 25th, 50th, and 75th percentiles of prices.

**Hint:** np.percentile()

In [ ]:
# Your code here


### Question 47: Quantiles (Easy)

Calculate quartiles of quantity data.

**Hint:** np.quantile()

In [ ]:
# Your code here


### Question 48: Sum of Revenue (Easy)

Calculate total revenue across all transactions.

**Hint:** Sum of quantity * price

In [ ]:
# Your code here


### Question 49: Count Non-Zero (Easy)

Count how many non-zero quantities exist.

**Hint:** np.count_nonzero()

In [ ]:
# Your code here


### Question 50: Average Revenue per Transaction (Easy)

Calculate mean transaction value.

**Hint:** Mean of revenue array

In [ ]:
# Your code here


### Question 51: Price Distribution (Easy)

Create a histogram of prices (get counts and bins).

**Hint:** np.histogram()

In [ ]:
# Your code here


### Question 52: Correlation Coefficient (Easy)

Calculate correlation between quantity and price.

**Hint:** np.corrcoef()

In [ ]:
# Your code here


### Question 53: Covariance (Easy)

Calculate covariance between quantity and price.

**Hint:** np.cov()

In [ ]:
# Your code here


### Question 54: Unique Quantities (Easy)

Find all unique quantity values.

**Hint:** np.unique()

In [ ]:
# Your code here


### Question 55: Count Unique (Easy)

Count how many unique prices exist.

**Hint:** len(np.unique())

In [ ]:
# Your code here


### Question 56: Mode Approximation (Easy)

Find the most common quantity value.

**Hint:** Use unique with return_counts

In [ ]:
# Your code here


### Question 57: Range Calculation (Easy)

Calculate the range of quantities (max - min).

**Hint:** np.ptp()

In [ ]:
# Your code here


### Question 58: Sum by Axis (Easy)

For a 2D subset, calculate sum along rows and columns.

**Hint:** np.sum(axis=0) or axis=1

In [ ]:
# Your code here


### Question 59: Mean by Axis (Easy)

Calculate mean along specific axis for subset.

**Hint:** np.mean(axis=)

In [ ]:
# Your code here


### Question 60: Weighted Average (Easy)

Calculate weighted average price by quantity.

**Hint:** np.average() with weights

In [ ]:
# Your code here


## Section 1.4: Indexing and Slicing (Questions 61-80)

### Question 61: Single Element (Easy)

Access the 100th transaction's unit price.

**Hint:** Use index [99]

In [ ]:
# Your code here


### Question 62: Row Extraction (Easy)

Extract the complete 50th transaction.

**Hint:** Use row indexing

In [ ]:
# Your code here


### Question 63: Column Extraction (Easy)

Extract all values from the quantity column.

**Hint:** Use column indexing

In [ ]:
# Your code here


### Question 64: Subset Selection (Easy)

Get transactions 100-200.

**Hint:** Use slicing [100:200]

In [ ]:
# Your code here


### Question 65: Multiple Columns (Easy)

Extract quantity, unit_price, and total_amount columns.

**Hint:** Use array indexing

In [ ]:
# Your code here


### Question 66: Every 5th Row (Easy)

Select every 5th transaction.

**Hint:** Use step in slicing [::5]

In [ ]:
# Your code here


### Question 67: Reverse Array (Easy)

Reverse the order of transactions.

**Hint:** Use [::-1]

In [ ]:
# Your code here


### Question 68: Boolean Indexing (Easy)

Select transactions where quantity > 5.

**Hint:** Use boolean mask

In [ ]:
# Your code here


### Question 69: Price Range Filter (Easy)

Get transactions with price between $50 and $100.

**Hint:** Use compound boolean condition

In [ ]:
# Your code here


### Question 70: Multiple Conditions (Easy)

Select transactions where quantity > 3 AND price < 100.

**Hint:** Use & operator

In [ ]:
# Your code here


### Question 71: OR Condition (Easy)

Select where quantity = 1 OR quantity = 10.

**Hint:** Use | operator

In [ ]:
# Your code here


### Question 72: NOT Condition (Easy)

Select transactions where quantity is NOT equal to 5.

**Hint:** Use ~ operator

In [ ]:
# Your code here


### Question 73: Fancy Indexing (Easy)

Select transactions at indices [0, 5, 10, 15, 20].

**Hint:** Use array of indices

In [ ]:
# Your code here


### Question 74: Top 10 Prices (Easy)

Get indices of top 10 highest prices.

**Hint:** Use np.argsort()[-10:]

In [ ]:
# Your code here


### Question 75: Bottom 5 Quantities (Easy)

Get indices of 5 smallest quantities.

**Hint:** Use np.argsort()[:5]

In [ ]:
# Your code here


### Question 76: Where Function (Easy)

Find indices where price > 200.

**Hint:** Use np.where()

In [ ]:
# Your code here


### Question 77: Replace Values (Easy)

Replace all quantities of 1 with 2.

**Hint:** Use boolean indexing assignment

In [ ]:
# Your code here


### Question 78: Clip Values (Easy)

Clip quantities between 2 and 8.

**Hint:** Use np.clip()

In [ ]:
# Your code here


### Question 79: Select by Index List (Easy)

Select specific transactions by their IDs.

**Hint:** Use fancy indexing

In [ ]:
# Your code here


### Question 80: 2D Indexing (Easy)

From a 2D subset, select specific rows and columns.

**Hint:** Use [rows, cols]

In [ ]:
# Your code here


## Section 1.5: Array Manipulation (Questions 81-100)

### Question 81: Append Values (Easy)

Append new quantity value to array.

**Hint:** Use np.append()

In [ ]:
# Your code here


### Question 82: Insert Value (Easy)

Insert a value at specific position.

**Hint:** Use np.insert()

In [ ]:
# Your code here


### Question 83: Delete Values (Easy)

Delete first 10 elements from quantity array.

**Hint:** Use np.delete()

In [ ]:
# Your code here


### Question 84: Concatenate Arrays (Easy)

Combine quantity and price arrays.

**Hint:** Use np.concatenate()

In [ ]:
# Your code here


### Question 85: Vertical Stack (Easy)

Stack multiple transaction rows vertically.

**Hint:** Use np.vstack()

In [ ]:
# Your code here


### Question 86: Horizontal Stack (Easy)

Stack columns horizontally.

**Hint:** Use np.hstack()

In [ ]:
# Your code here


### Question 87: Split Array (Easy)

Split quantity array into 10 equal parts.

**Hint:** Use np.split()

In [3]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
print(np.split(df['quantity'],10))


[0      9
1      3
2      2
3      4
4      4
      ..
995    6
996    8
997    5
998    2
999    9
Name: quantity, Length: 1000, dtype: int64, 1000    4
1001    4
1002    7
1003    6
1004    4
       ..
1995    9
1996    7
1997    9
1998    3
1999    9
Name: quantity, Length: 1000, dtype: int64, 2000    2
2001    9
2002    6
2003    4
2004    2
       ..
2995    3
2996    9
2997    7
2998    9
2999    2
Name: quantity, Length: 1000, dtype: int64, 3000    6
3001    5
3002    2
3003    4
3004    4
       ..
3995    8
3996    8
3997    3
3998    8
3999    4
Name: quantity, Length: 1000, dtype: int64, 4000    3
4001    9
4002    8
4003    2
4004    8
       ..
4995    8
4996    7
4997    9
4998    5
4999    1
Name: quantity, Length: 1000, dtype: int64, 5000    8
5001    2
5002    6
5003    1
5004    6
       ..
5995    7
5996    6
5997    6
5998    8
5999    8
Name: quantity, Length: 1000, dtype: int64, 6000    5
6001    4
6002    5
6003    9
6004    6
       ..
6995    6
6996    7
6997  

c:\ProgramData\anaconda3\Lib\site-packages\numpy\_core\fromnumeric.py:57: FutureWarning: 'Series.swapaxes' is deprecated and will be removed in a future version. Please use 'Series.transpose' instead.
  return bound(*args, **kwds)


### Question 88: Array Split (Easy)

Split at specific indices.

**Hint:** Use np.array_split()

In [ ]:
# Your code here


### Question 89: Tile Array (Easy)

Repeat quantity array 3 times.

**Hint:** Use np.tile()

In [5]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
print(np.tile(df['quantity'],3))

[9 3 2 ... 9 3 3]


### Question 90: Repeat Elements (Easy)

Repeat each quantity value twice.

**Hint:** Use np.repeat()

In [ ]:
# Your code here


### Question 91: Unique with Counts (Easy)

Get unique quantities and their counts.

**Hint:** np.unique(return_counts=True)

In [7]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
values, counts = np.unique(df['quantity'], return_counts=True)
print(values)
print(counts)

[1 2 3 4 5 6 7 8 9]
[1028 1059 1133 1106 1193 1122 1134 1090 1135]


### Question 92: Sort Array (Easy)

Sort quantities in ascending order.

**Hint:** Use np.sort()

In [8]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
print(np.sort(df['quantity']))


[1 1 1 ... 9 9 9]


### Question 93: Argsort (Easy)

Get indices that would sort prices.

**Hint:** Use np.argsort()

In [12]:
# Your code here
import pandas as pd
import numpy as np
df=pd.read_csv("retail_data.csv")
df
print(np.argsort(df['unit_price']))

0       5268
1       2393
2       5687
3       7626
4        750
        ... 
9995    6506
9996    8523
9997    2817
9998    1423
9999    5283
Name: unit_price, Length: 10000, dtype: int64


### Question 94: Reverse Sort (Easy)

Sort prices in descending order.

**Hint:** Use np.sort()[::-1]

In [15]:
# Your code here
print(np.sort(df['unit_price'][::-1]))

[  5.     5.03   5.05 ... 499.85 499.87 499.9 ]


### Question 95: Searchsorted (Easy)

Find position to insert value in sorted array.

**Hint:** Use np.searchsorted()

In [ ]:
# Your code here


### Question 96: Roll Array (Easy)

Roll/shift quantity array by 5 positions.

**Hint:** Use np.roll()

In [ ]:
# Your code here


### Question 97: Flip Array (Easy)

Flip array upside down.

**Hint:** Use np.flip()

In [ ]:
# Your code here


### Question 98: Rot90 (Easy)

Rotate 2D subset by 90 degrees.

**Hint:** Use np.rot90()

In [ ]:
# Your code here


### Question 99: Trim Zeros (Easy)

Remove leading/trailing zeros if any.

**Hint:** Use np.trim_zeros()

In [ ]:
# Your code here


### Question 100: Pad Array (Easy)

Add zeros to beginning and end of array.

**Hint:** Use np.pad()

In [ ]:
# Your code here


---

# PART 2: INTERMEDIATE LEVEL (100 Questions)

## Section 2.1: Advanced Array Operations (Questions 101-120)

### Question 101: Broadcasting Addition (Intermediate)

Add a 1D array to each row of 2D price matrix.

**Hint:** Use broadcasting rules

In [ ]:
# Your code here


### Question 102: Broadcasting Multiplication (Intermediate)

Multiply 2D array by column vector.

**Hint:** Reshape for broadcasting

In [ ]:
# Your code here


### Question 103: Outer Product (Intermediate)

Calculate outer product of quantity and price arrays (first 100).

**Hint:** Use np.outer()

In [ ]:
# Your code here


### Question 104: Dot Product (Intermediate)

Calculate dot product of two transaction vectors.

**Hint:** Use np.dot()

In [ ]:
# Your code here


### Question 105: Matrix Multiplication (Intermediate)

Perform matrix multiplication on subsets.

**Hint:** Use np.matmul() or @

In [ ]:
# Your code here


### Question 106: Einstein Sum (Intermediate)

Use einsum for custom array operations.

**Hint:** Use np.einsum()

In [ ]:
# Your code here


### Question 107: Meshgrid (Intermediate)

Create 2D grid for quantity and price combinations.

**Hint:** Use np.meshgrid()

In [ ]:
# Your code here


### Question 108: Gradient (Intermediate)

Calculate gradient of price array.

**Hint:** Use np.gradient()

In [ ]:
# Your code here


### Question 109: Diff Operation (Intermediate)

Calculate differences between consecutive prices.

**Hint:** Use np.diff()

In [ ]:
# Your code here


### Question 110: Convolve (Intermediate)

Apply moving average using convolution.

**Hint:** Use np.convolve()

In [ ]:
# Your code here


### Question 111: Moving Average (Intermediate)

Calculate 7-day moving average of quantities.

**Hint:** Use window convolution

In [ ]:
# Your code here


### Question 112: Piecewise (Intermediate)

Create piecewise function on prices.

**Hint:** Use np.piecewise()

In [ ]:
# Your code here


### Question 113: Vectorize Function (Intermediate)

Vectorize custom discount calculation function.

**Hint:** Use np.vectorize()

In [ ]:
# Your code here


### Question 114: Apply Along Axis (Intermediate)

Apply custom function along specific axis.

**Hint:** Use np.apply_along_axis()

In [ ]:
# Your code here


### Question 115: Polynomial Fit (Intermediate)

Fit polynomial to quantity vs price data.

**Hint:** Use np.polyfit()

In [ ]:
# Your code here


### Question 116: Polynomial Evaluation (Intermediate)

Evaluate fitted polynomial.

**Hint:** Use np.polyval()

In [ ]:
# Your code here


### Question 117: Polynomial Roots (Intermediate)

Find roots of polynomial equation.

**Hint:** Use np.roots()

In [ ]:
# Your code here


### Question 118: Interpolation (Intermediate)

Interpolate missing price values.

**Hint:** Use np.interp()

In [ ]:
# Your code here


### Question 119: Digital Filter (Intermediate)

Apply digital filter to time series data.

**Hint:** Use convolution or scipy

In [ ]:
# Your code here


### Question 120: FFT (Intermediate)

Perform Fast Fourier Transform on quantity data.

**Hint:** Use np.fft.fft()

In [ ]:
# Your code here


## Section 2.2: Aggregations and GroupBy Operations (Questions 121-140)

### Question 121: Calculate revenue by customer_id groups (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 122: Find average quantity per store_id (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 123: Group by payment method and sum (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 124: Calculate daily revenue totals (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 125: Find top 10 customers by revenue (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 126: Calculate hourly sales patterns (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 127: Group by day of week (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 128: Calculate moving totals (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 129: Find seasonal patterns (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 130: Calculate growth rates (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 131: Identify outliers per group (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 132: Calculate percentiles by category (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 133: Compute weighted metrics by group (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 134: Find mode per categorical variable (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 135: Calculate variance by segment (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 136: Create pivot table structure (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 137: Cross-tabulation with NumPy (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 138: Calculate cohort analysis data (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 139: Time-based aggregations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 140: Multi-level aggregations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


## Section 2.3: Linear Algebra Operations (Questions 141-160)

### Question 141: Matrix inverse of subset (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 142: Calculate determinant (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 143: Solve linear equations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 144: Eigenvalues and eigenvectors (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 145: Matrix rank calculation (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 146: QR decomposition (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 147: SVD decomposition (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 148: Cholesky decomposition (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 149: Matrix norm calculations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 150: Condition number (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 151: Pseudo-inverse calculation (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 152: Matrix exponential (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 153: Matrix power (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 154: Kronecker product (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 155: Trace calculation (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 156: Diagonal matrix operations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 157: Triangular matrices (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 158: Orthogonal matrices (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 159: Symmetric matrices (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 160: Positive definite check (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


## Section 2.4: Random Operations and Sampling (Questions 161-180)

### Question 161: Random sampling without replacement (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 162: Stratified sampling (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 163: Bootstrap sampling (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 164: Random permutations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 165: Shuffle transactions (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 166: Random choice with probabilities (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 167: Generate random prices (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 168: Simulate customer arrivals (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 169: Monte Carlo simulation (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 170: Random walk generation (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 171: Binomial distribution sampling (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 172: Poisson distribution for arrivals (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 173: Normal distribution generation (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 174: Exponential distribution (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 175: Uniform random generation (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 176: Random seed for reproducibility (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 177: Generate correlated variables (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 178: Random matrix generation (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 179: Sampling from custom distribution (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 180: Cross-validation splits (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


## Section 2.5: Advanced Indexing and Masking (Questions 181-200)

### Question 181: Advanced boolean indexing (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 182: Multi-dimensional indexing (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 183: Index arrays with arrays (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 184: Complex slicing operations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 185: Masked array operations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 186: Conditional selections (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 187: Index manipulation (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 188: Advanced where operations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 189: Take along axis (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 190: Put operations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 191: Choose function usage (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 192: Extract diagonal elements (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 193: Diagonal indexing (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 194: Structured array indexing (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 195: Record array operations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 196: Field access in structured arrays (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 197: View vs copy operations (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 198: Memory-efficient indexing (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 199: Index broadcasting (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


### Question 200: Advanced fancy indexing patterns (Intermediate)

Implement this operation using NumPy on the retail dataset.

**Hint:** Research the specific NumPy functions needed for this operation.

In [ ]:
# Your code here


---

# PART 3: ADVANCED LEVEL (100 Questions)

## Section 3.1: Performance Optimization (Questions 201-220)

### Question 201: Vectorization vs Loops (Advanced)

Compare performance of vectorized vs loop operations.

**Hint:** Use timeit

In [ ]:
# Your code here


### Question 202: Memory Optimization (Advanced)

Optimize memory usage with appropriate dtypes.

**Hint:** Use astype()

In [ ]:
# Your code here


### Question 203: View vs Copy Performance (Advanced)

Demonstrate performance difference.

**Hint:** Measure time and memory

In [ ]:
# Your code here


### Question 204: In-place Operations (Advanced)

Use in-place operations to save memory.

**Hint:** Use +=, *=, etc.

In [ ]:
# Your code here


### Question 205: Broadcasting Optimization (Advanced)

Optimize using broadcasting instead of loops.

**Hint:** Eliminate explicit loops

In [ ]:
# Your code here


### Question 206: Stride Tricks (Advanced)

Use stride_tricks for efficient windowing.

**Hint:** np.lib.stride_tricks

In [ ]:
# Your code here


### Question 207: Numba Integration (Advanced)

Speed up with Numba JIT compilation.

**Hint:** Use @njit decorator

In [ ]:
# Your code here


### Question 208: Parallel Processing (Advanced)

Implement parallel computations.

**Hint:** Use multiprocessing

In [ ]:
# Your code here


### Question 209: Memory Mapping (Advanced)

Work with large datasets using memmap.

**Hint:** np.memmap()

In [ ]:
# Your code here


### Question 210: Lazy Evaluation (Advanced)

Implement lazy evaluation patterns.

**Hint:** Use generators

In [ ]:
# Your code here


### Question 211: Cache Optimization (Advanced)

Optimize cache usage in calculations.

**Hint:** Consider data locality

In [ ]:
# Your code here


### Question 212: SIMD Operations (Advanced)

Leverage SIMD instructions.

**Hint:** Ensure proper alignment

In [ ]:
# Your code here


### Question 213: Reduce Memory Footprint (Advanced)

Minimize memory for large arrays.

**Hint:** Use compression

In [ ]:
# Your code here


### Question 214: Efficient Aggregations (Advanced)

Optimize group-by style operations.

**Hint:** Use ufunc.reduceat

In [ ]:
# Your code here


### Question 215: Fast Searching (Advanced)

Implement efficient search algorithms.

**Hint:** Use searchsorted

In [ ]:
# Your code here


### Question 216: Optimized Sorting (Advanced)

Use most efficient sorting approach.

**Hint:** Choose right algorithm

In [ ]:
# Your code here


### Question 217: Batch Processing (Advanced)

Process data in efficient batches.

**Hint:** Chunk processing

In [ ]:
# Your code here


### Question 218: Avoid Copies (Advanced)

Minimize unnecessary array copies.

**Hint:** Use views where possible

In [ ]:
# Your code here


### Question 219: Efficient Concatenation (Advanced)

Optimize array concatenation.

**Hint:** Pre-allocate when possible

In [ ]:
# Your code here


### Question 220: Profile Code (Advanced)

Profile NumPy code for bottlenecks.

**Hint:** Use line_profiler

In [ ]:
# Your code here


## Section 3.2: Complex Statistical Analysis (Questions 221-240)

### Question 221: Multivariate statistics (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 222: PCA implementation (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 223: Correlation matrices (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 224: Hypothesis testing (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 225: Chi-square tests (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 226: ANOVA implementation (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 227: Time series decomposition (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 228: Autocorrelation (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 229: Cross-correlation (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 230: Rolling statistics (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 231: Exponential smoothing (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 232: Trend analysis (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 233: Seasonality detection (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 234: Outlier detection algorithms (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 235: Anomaly detection (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 236: Distribution fitting (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 237: Goodness of fit tests (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 238: Bootstrap methods (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 239: Confidence intervals (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 240: Resampling techniques (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


## Section 3.3: Advanced Linear Algebra (Questions 241-260)

### Question 241: Sparse matrix operations (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 242: Matrix factorizations (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 243: Iterative solvers (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 244: Least squares optimization (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 245: Convex optimization (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 246: Constrained optimization (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 247: Numerical integration (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 248: Differential equations (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 249: Root finding (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 250: Interpolation splines (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 251: Curve fitting (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 252: Function minimization (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 253: Gradient descent (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 254: Conjugate gradients (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 255: Newton methods (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 256: Trust region methods (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 257: Linear programming (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 258: Quadratic programming (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 259: Eigenvalue problems (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 260: Generalized eigenvalue problems (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


## Section 3.4: Signal Processing and Transforms (Questions 261-280)

### Question 261: Fourier transforms (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 262: Inverse Fourier transforms (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 263: Power spectral density (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 264: Spectral analysis (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 265: Wavelet transforms (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 266: Filter design (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 267: Digital signal processing (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 268: Convolution theorem (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 269: Window functions (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 270: Frequency domain analysis (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 271: Phase analysis (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 272: Amplitude spectrum (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 273: Coherence analysis (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 274: Cross-spectral density (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 275: Transfer functions (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 276: System identification (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 277: Deconvolution (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 278: Hilbert transform (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 279: Envelope detection (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 280: Time-frequency analysis (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


## Section 3.5: Advanced Techniques and Algorithms (Questions 281-300)

### Question 281: Custom ufuncs (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 282: Advanced broadcasting (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 283: Structured arrays complex (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 284: Record arrays advanced (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 285: Memory layout optimization (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 286: Custom dtypes (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 287: Subclassing ndarray (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 288: Array interface (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 289: Array protocol (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 290: Masked array advanced (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 291: Matrix class usage (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 292: Polynomial class (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 293: Random number generators (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 294: Bit generators (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 295: Statistical distributions (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 296: Advanced indexing tricks (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 297: Tensor operations (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 298: N-dimensional algorithms (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 299: Advanced array manipulation (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


### Question 300: Real-world optimization challenge (Advanced)

Implement this advanced NumPy technique using the retail dataset.

**Hint:** This requires deep NumPy knowledge. Consult documentation and research papers.

In [ ]:
# Your code here


---

## Conclusion

Congratulations on completing all 300 NumPy questions!

### Key Takeaways:
- ✅ Mastered NumPy array operations
- ✅ Learned vectorization techniques
- ✅ Optimized code performance
- ✅ Applied linear algebra concepts
- ✅ Worked with real-world retail data

### Next Steps:
1. Review any challenging questions
2. Optimize your solutions
3. Compare with NumPy documentation examples
4. Apply these skills to your own datasets

---

**Total Questions Completed: 300**
**Skill Level Achieved: NumPy Expert 🎓**